File: src/Runtime/Sentence/SentenceReturn.cs
/*
	PeakScript: https://github.com/Jenocn/PeakScript
	By Jenocn: https://jenocn.github.io
*/

namespace peak.interpreter {

	public class SentenceReturn : Sentence {
		private SentenceExpression _returnExpression = null;
		private Value _returnValue = null;

		public Value returnValue { get => _returnValue; }

		public SentenceReturn() { }

		public SentenceReturn(SentenceExpression expression) {
			_returnExpression = expression;
		}
		public override ExecuteResult Execute(Space space) {
			if (!IsSuccess(_returnExpression.Execute(space))) {
				ErrorLogger.LogRuntimeError(ErrorRuntimeCode.Return, "The expression execute failed!");
				return ExecuteResult.Failed;
			}
			SetReturnValue(_returnExpression.value);
			return ExecuteResult.Return;
		}

		protected void SetReturnValue(Value value) {
			_returnValue = value;
		}

	}
} // namespace peak.interpreter

File: src/runtime/sentence/sentence_return.cpp
#include "sentence_return.h"
#include "sentence_expression.h"
#include "runtime/value/value_tool.h"

using namespace peak;

SentenceReturn::SentenceReturn(std::shared_ptr<SentenceExpression> expression)
	: _returnExpression(expression) {
}
ExecuteResult SentenceReturn::Execute(std::shared_ptr<Space> space) {
	if (_returnExpression) {
		if (!IsSuccess(_returnExpression->Execute(space))) {
			ErrorLogger::LogRuntimeError(ErrorRuntimeCode::Return, "The expression execute failed!");
			return ExecuteResult::Failed;
		}
		SetReturnValue(_returnExpression->GetValue());
	} else {
		SetReturnValue(ValueNull::DEFAULT_VALUE);
	}
	return ExecuteResult::Return;
}

std::shared_ptr<Value> SentenceReturn::GetReturnValue() const {
	return _returnValue;
}

void SentenceReturn::SetReturnValue(std::shared_ptr<Value> value) {
	_returnValue = value;
}
